### PR TITLE
Feature/f tmp leg uk

### PR DESCRIPTION
**Pull request by TheMasteredPanda**
At the request of FlyingP1g I am producing this request to merge what I have written currently.

Includes the following commands:
  - bills search
  - bills info
  - divisions csearch
  - divisions cinfo
  - divisions linfo
  - divisions lsearch

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -35,17 +38,36 @@ def __init__(self):
...
 
-        self.google_drive = modules.google_drive.Drive()
+#        self.google_drive = modules.google_drive.Drive()
         self.timers = modules.timers.Timers(self)
         self.reaction_menus = modules.reaction_menus.ReactionMenus(self)
         self.custom_commands = modules.custom_commands.CustomCommands(self)
         self.leveling_system = modules.leveling.LevelingSystem(self)
+        self.ukparl_module = modules.ukparliament.UKParliamentModule(self.ukparliament)
+
+    async def on_ready(self):
```
Comment (Hattyot) [inline bot.py:59]: the default on_ready should not be overwritten. You should move it's functionality to it's own function in the Events cog and call it in Events on_ready listener.

File: bot.py [hunk 2]
```diff
@@ -35,17 +38,36 @@ def __init__(self):
...
+        guild: Guild = self.guilds[0]
+        channel = guild.get_channel(config.UKPARL_CHANNEL)
+        self.ukparl_module.set_guild(self.guilds[0])
+        if self.ukparl_module.tracker_event_loop.is_running() is False:
+            self.ukparl_module.tracker_event_loop.start()
+
+    def get_parliament(self):
```
Comment (Hattyot) [inline bot.py:66]: this function should be removed, i dont think it's needed

File: bot.py [hunk 3]
```diff
@@ -35,17 +38,36 @@ def __init__(self):
...
+        return self.ukparliament
+
+    def get_parliament_module(self):
```
Comment (Hattyot) [inline bot.py:69]: this function should be removed, i dont think it's needed

File: modules/ukparliament.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        self.guild: Union[Guild, None] = None
+        if os.path.exists('tmpimages') is False:
+            os.mkdir('tmpimages')
+
+    def set_guild(self, guild):
+        self.guild = guild
+
+    @tasks.loop(seconds=60)
+    async def tracker_event_loop(self):
+        await self.parliament.get_bills_tracker()._poll()
```
Comment (Hattyot) [inline modules/ukparliament.py:162]: if you're gonna access values of an outside class, like here, I dont think you should make it a protected value

File: cogs/uk_parliament.py [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+from ukparliament.structures.bills import Bill, BillType, CommonsDivision, LordsDivision
+from bot import TLDR
+from modules import cls, embed_maker
+from modules.utils import ParseArgs, get_custom_emote, get_member_from_string
+from discord.ext import commands, tasks
+from discord.ext.commands import Context
+from ukparliament.bills import SearchBillsBuilder, SearchBillsSortOrder
+from collections import namedtuple
+
+Pair = namedtuple("Pair", ["enum", "max_page_num"])
```
Comment (Hattyot) [inline cogs/uk_parliament.py:23]: what's this used for?

File: cogs/uk_parliament.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+            await embed_maker.message(ctx, description=f"Couldn't find latest elections results for {'Borough' if args['borough'] != '' and args['borough'] is not None else 'MP'} {args['borough'] if args['borough'] != '' and args['borough'] is not None else args['pre']}", send=True)
+            return
+
+    
+
+        next_line = '\n'
+        results = await self.parliament.get_election_results(member)
+        result = results[0]
+
+        if args['historical'] is '':
```
Comment (Hattyot) [inline cogs/uk_parliament.py:217]: should change the is to ==

File: cogs/uk_parliament.py [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+    async def division_lord_info(self, ctx: commands.Context, division_id: int):
+        division = await self.parliament.get_lords_division(division_id)
+        if division is None:
+            await embed_maker.message(ctx, description=f"Couldn't find division under id {division_id}", send=True)
+
+
+        image_file = self.bot.get_parliament_module().generate_division_image(self.parliament, division)
+        next_line = '\n'
+        did_pass = division.get_aye_count() > division.get_no_count()
+        embed: embeds.Embed = await embed_maker.message(ctx, description=f"**Title:** {division.get_division_title()}{next_line}**Division Outcome:** {'Passed' if did_pass else 'Not passed'} by a division of {division.get_aye_count() if did_pass else division.get_no_count()} {'Ayes' if did_pass else 'Noes'} to {division.get_no_count() if did_pass else division.get_aye_count()} {'Noes' if did_pass else 'Ayes'}{next_line}**Division Date:** {division.get_division_date().strftime('%Y-%m-%d %H:%M:%S')}{next_line}**Summary:** {division.get_amendment_motion_notes()[0:250]}") #type: ignore
```
Comment (Hattyot) [inline cogs/uk_parliament.py:335]: This def needs to be split into a bunch of different lines

File: modules/ukparliament.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+    async def on_lords_division(self, division: LordsDivision, bill: Bill):
+        channel = self.guild.get_channel(config.LORDS_CHANNEL)
+        division_infographic = self.generate_division_image(self.parliament, division)
+        embed = Embed(color=discord.Colour.from_rgb(166, 42, 22), timestamp=datetime.now())
+        did_pass = division.get_aye_count() > division.get_no_count()
+        embed.title = f'**{division.get_division_title()}**'
+        embed.set_image(url=f"attachment://{division_infographic.filename}")
+        next_line = '\n'
+        description = f"**ID:** {division.get_id()}{next_line}**Summary [(Link)](https://votes.parliament.uk/Votes/Lords/Division/{division.get_id()}):** {division.get_amendment_motion_notes()[0:250]}{next_line}**Division Result:** {'Passed' if did_pass else 'Not passed'} by a division of {division.get_aye_count() if did_pass else division.get_no_count()} {'Ayes' if did_pass else 'Noes'} to {division.get_no_count() if did_pass else division.get_aye_count()} {'Noes' if did_pass else 'Ayes'}{next_line}**Division Date:** {division.get_division_date().strftime('%Y-%m-%d %H:%M:%S')}"
```
Comment (Hattyot) [inline modules/ukparliament.py:212]: This also really needs to be split into separate lines.

**Review by Hattyot — CHANGES_REQUESTED** [review]
More generic initial notes:

1. add back config_example.py to keep track of all the stuff that needs to be in the config

2. update requirements with the new modules added to the bot, like matplotlib

3. Try to comply more with pep8 rules, except the line length one, that can mostly be ignored.

4. instead of being static i think it would be nice if config.LORDS_CHANNEL and other values like that could be set through a command, that part can be delayed for later though

5. try to keep track of unused variables and remove them

6. for the ukparliament please provide documentation like in the other modules

7. this is a minor thing, but sort the imports by "import x" at the top and "from y import x" at the bottom sorted by an enter, if not possible and it causes conflicts or something ignore this

8. move from json storage to mongodb

9. For lines that do get exceedingly long, like line 420 in uk_parliament cog, it would be nice to split the code into sections on seprate lines.

10. instead of leaving commands empty with a "pass" like divisions command in uk_parliament cog. Add "return await embed_maker.command_error(ctx)". It'll return help data on the command.

11. also keep track of unused imports

12. try to avoid "is" with a literal. like line 217 in uk_parliament cog

13. UKCommand class can be renamed to just UK.

14. If args in commands is None, also return "await embed_maker.command_error(ctx)" to send the user info about the command.

15. when possible, instead of establishing a new aiohttp client session, grab it from the bot "aiohttp_session = getattr(self.bot.http, '_HTTPClient__session')"

16. In the embeds, i think the "ID" value can be left out

17. This is something we should establish together, but the image creation will need to be offloaded to a separate api hosted on heroku or smth.

18. For the command_args in all the different commands, i think you should still add add a short version of the arg and provide an example for every command_arg and limits if needed.

File: bot.py (hunk 2 above)
Comment (TheMasteredPanda) [reply bot.py]: Used in the the modules script of this feature.

File: modules/ukparliament.py (hunk 4 above)
Comment (TheMasteredPanda) [reply modules/ukparliament.py:112]: Fixed.

File: cogs/uk_parliament.py (hunk 6 above)
Comment (TheMasteredPanda) [reply cogs/uk_parliament.py]: Fixed.

File: cogs/uk_parliament.py (hunk 7 above)
Comment (TheMasteredPanda) [reply cogs/uk_parliament.py]: Done.

File: modules/ukparliament.py (hunk 8 above)
Comment (TheMasteredPanda) [reply modules/ukparliament.py]: Again. fixed.

**Pull request closed without merging**